src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1956,6 +1968,568 @@ public Task OpenAsync(CancellationToken token)
+                try
+                {
+                    if (!TryReceiveRequest(timeoutHelper.RemainingTime(), out context))
+                    {
+                        return (false, wasAborted);
+                    }
+
+                    if (context != null && context.RequestMessage !=null)
+                    {
+                        Message message = context.RequestMessage;
+                        using (message)
+                        {
+                            ProtocolException error = ProtocolException.ReceiveShutdownReturnedNonNull(message);
+                            throw TraceUtility.ThrowHelperWarning(error, message);
+                        }
+                    }
+                    bool result = await _inputSessionCloseHandle.WaitAsync(timeoutHelper.RemainingTime(), false);
+                    if (!result)
+                    {
+                        return (false, wasAborted);
+                    }
+                    else
+                    {
+                        lock (LocalLock)
+                        {
+                            if (!(_isInputClosed))
+                            {
+                                Fx.Assert("Shutdown request was not received.");
+                                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.ShutdownRequestWasNotReceived)));
+                            }
+                        }
+                        return (true, wasAborted);
+                    }
+                }
+                catch (CommunicationObjectAbortedException)
+                {
+                    if (State != CommunicationState.Closed)
+                    {
+                        throw;
+                    }
+                    wasAborted = true;
+                }
+                return (false, wasAborted);
+            }
+
+            private class SoapSecurityServerDuplexSession : SoapSecurityInputSession, IDuplexSession
+            {
+                private ServerSecurityDuplexSessionChannel _channel;
+
+                public SoapSecurityServerDuplexSession(SecurityContextSecurityToken sessionToken, SecuritySessionServerSettings settings, ServerSecurityDuplexSessionChannel channel)
+                    : base(sessionToken, settings, channel)
+                {
+                    _channel = channel;
+                }
+
+                public void CloseOutputSession()
+                {
+                    CloseOutputSession(ServiceDefaults.CloseTimeout);
+                }
+
+                public void CloseOutputSession(TimeSpan timeout)
+                {
+                   // channel.ThrowIfFaulted();
+                   // channel.ThrowIfNotOpened();
+                    Exception pendingException = null;
+                    try
+                    {
+                        _channel.CloseOutputSessionAsync(new TimeoutHelper(timeout).GetCancellationToken());
+                    }
```

Comment:
This method call needs to be awaited otherwise some exceptions won't get thrown and handled here. Which means this method needs to be async.